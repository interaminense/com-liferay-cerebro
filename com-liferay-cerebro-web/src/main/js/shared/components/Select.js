import Component, {Config} from 'metal-jsx';

class SelectItem extends Component {
	render() {
		return <option {...this.otherProps()}>{this.props.children}</option>;
	}
}

class Select extends Component {
	render() {
		const {children, showBlankOption} = this.props;

		return (
			<select class="form-control select-root" {...this.otherProps()}>
				{showBlankOption && <SelectItem />}

				{children}
			</select>
		);
	}
}

Select.PROPS = {
	showBlankOption: Config.bool().value(false)
};

Select.Item = SelectItem;

export default Select;