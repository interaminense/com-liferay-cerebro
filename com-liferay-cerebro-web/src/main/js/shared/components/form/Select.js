import autobind from 'autobind-decorator';
import Component, {Config} from './Component';
import getCN from 'classnames';
import {noop, uniqueId} from 'lodash';

import Label from './Label';
import Select from '../Select';
import validate, {checkRequired} from '../../util/form-validators';

class FormSelect extends Component {
	created() {
		this._id = uniqueId('FormSelect');
	}

	attached() {
		super.attached();

		this.props.children.forEach(child => {
			const {selected, value} = child.props;

			if (selected) {
				this.state.value = value;
			}
		});
	}

	@autobind
	handleChange(event) {
		this.state.value = event.target.value;

		if (this.state.show) {
			this.validateFormElement();
		}

		this.props.onChange(event.target.value);
	}

	isRequired() {
		const {validator} = this.props;

		return validator && checkRequired(validator);
	}

	validateFormElement(showErrors = true) {
		const {validator} = this.props;

		let retVal = true;

		if (validator) {
			const result = validate(this.state.value, validator).find(
				item => !item.valid
			);

			const valid = result ? result.valid : true;

			this.setState({
				message: result ? result.message : '',
				show: showErrors,
				valid
			});

			retVal = valid;
		}

		return retVal;
	}

	render() {
		const {
			props: {children, info, inline, label},
			state: {message, show, valid, value}
		} = this;

		const classes = getCN({
			'form-inline-group': inline,
			'has-danger': !valid,
			'has-success': valid && show
		});

		return (
			<div class={classes}>
				{label && (
					<Label
						for={this._id}
						info={info}
						required={this.isRequired()}
					>
						{label}
					</Label>
				)}

				<Select
					{...this.otherProps()}
					id={this._id}
					onChange={this.handleChange}
					value={value}
				>
					{children}
				</Select>

				{show && <div class="help-block">{message}</div>}
			</div>
		);
	}
}

FormSelect.PROPS = {
	info: Config.string(),
	initialValue: Config.string().value(''),
	inline: Config.bool().value(false),
	label: Config.string(),
	onChange: Config.func().value(noop),
	validator: Config.object()
};

FormSelect.STATE = {
	message: Config.value(''),
	show: Config.value(false),
	valid: Config.value(true),
	value: Config.value('')
};

FormSelect.Item = Select.Item;

export default FormSelect;