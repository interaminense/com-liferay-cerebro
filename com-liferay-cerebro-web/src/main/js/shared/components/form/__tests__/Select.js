import Select from '../Select';

describe('Select', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = new Select({name: 'test'});

		expect(component).toMatchSnapshot();
	});

	it('should render with a label', () => {
		component = new Select({label: 'foo', name: 'test'});

		expect(component).toMatchSnapshot();
	});

	it('should render help block with a message', () => {
		component = new Select({name: 'test'});

		component.setState({message: 'test', show: true});

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should set state.value to new target value', () => {
		const foo = 'foo';
		component = new Select({name: 'test'});

		expect(component.state.value).not.toBe(foo);

		component.handleChange({target: {value: foo}});

		jest.runAllTimers();

		expect(component.state.value).toBe(foo);
	});

	it('should call `validateFormElement` if state.show is true', () => {
		component = new Select({name: 'test'});

		component.validateFormElement = jest.fn();

		component.state.show = true;

		expect(component.validateFormElement).not.toHaveBeenCalled();

		component.handleChange({target: {value: 'foo'}});

		expect(component.validateFormElement).toHaveBeenCalled();
	});

	it('should check if `required` validator is present', () => {
		component = new Select({name: 'test'});

		expect(component.isRequired()).toBeFalsy();

		component.props.validator = {required: true};

		jest.runAllTimers();

		expect(component.isRequired()).toBe(true);
	});
});