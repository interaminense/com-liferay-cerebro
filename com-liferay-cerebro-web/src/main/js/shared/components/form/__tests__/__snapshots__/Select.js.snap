// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render 1`] = `
<div class="">
  <select id="FormSelect1" value="" class="form-control select-root"></select>
</div>
`;

exports[`Select should render help block with a message 1`] = `
<div class="has-success">
  <select id="FormSelect3" value="" class="form-control select-root"></select>
  <div class="help-block">test</div>
</div>
`;

exports[`Select should render with a label 1`] = `
<div class="">
  <label for="FormSelect2" class="form-control-label label-root">foo</label>
  <select id="FormSelect2" value="" class="form-control select-root"></select>
</div>
`;
